### PR TITLE
Feature/#12 huggingface PE and gpt2 tokenizer

### PR DESCRIPTION
**Pull request by soeque1**
# Pull Request
레파지토리에 기여해주셔서 감사드립니다.

해당 PR을 제출하기 전에 아래 사항이 완료되었는지 확인 부탁드립니다:
- [x] 작성한 코드가 어떤 에러나 경고없이 빌드가 되었나요?
- [x] 충분한 테스트를 수행하셨나요?

## 1. 해당 PR은 어떤 내용인가요?

- GPT2 Tokenizer
- LR Scheduler ReduceLROnPlateau
- history left padding
- learnable positional embedding

## 2. PR과 관련된 이슈가 있나요?
- 학습이 완벽하지는 않지만 향상되었습니다.

### CODE REVIEW

**Comment by soeque1** [comment]
@ssaru @aisolab 이 번 PR은 리뷰 후 merge하겠습니다.

**Review by ssaru — APPROVED** [review]
@soeque1 

LGTM!
리뷰 완료했습니다.
고생하셨습니다.

**Comment by ssaru** [comment]
@inmoonlight 지형님도 NLP쪽이니 리뷰해주시면 어떨까 싶은데,, 괜찮으실까요?

File: src/model/net.py [hunk 1]
```diff
@@ -64,32 +71,50 @@ def init_lstm(self):
...
+        self, src: List[tensor], hidden: tuple = None, _length: List = None
+    ):
+        embedded = self.emb(src).transpose(0, 1)
+        embedded = torch.nn.utils.rnn.pack_padded_sequence(
+            embedded, _length, batch_first=False, enforce_sorted=False
+        )
         if hidden is None:
             hidden = (
-                torch.zeros(self.n_layers, src.shape[-1], self.hidden_size).to(
+                torch.zeros(self.n_layers, src.shape[0], self.hidden_size).to(
```
Comment (inmoonlight) [inline src/model/net.py:83]: `src` 가 List 라서 shape 이 찍히지 않을 것 같습니다..!

File: src/model/net.py [hunk 2]
```diff
@@ -64,32 +71,50 @@ def init_lstm(self):
             elif "weight" in name:
                 nn.init.xavier_normal(param)
 
-    def forward_rnn(self, src: List[tensor], hidden: tuple = None):
-        embedded = self.emb(src)
+    def forward_rnn(
+        self, src: List[tensor], hidden: tuple = None, _length: List = None
+    ):
```
Comment (inmoonlight) [inline src/model/net.py:76]: src 가 어떤 shape 을 갖고 있는지 docstring 으로 언급해주면 보기에 편할 것 같습니다 ㅎㅎ

```suggestion
    def forward_rnn(
        self, src: List[tensor], hidden: tuple = None, _length: List = None
    ):
    """
    Args:
        src: (batch_size, ...)
        hidden: 
        _length: 
    """
```

Comment (soeque1) [reply src/model/net.py:76]: 감사합니다 ~ docstring 추가하겠습니당

File: src/model/net.py (hunk 1 above)
Comment (soeque1) [reply src/model/net.py:83]: src가 List가 아니라 tensor인거 같은데.. typehint를 실수 한 것 같습니당!
typehint 체크 혹시 mypy 써야하나요? torch도 잘 지원 되는지?

File: src/model/net.py (hunk 2 above)
Comment (soeque1) [reply src/model/net.py:76]: https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring

@aisolab @ssaru @inmoonlight 이거 써볼까 하는데 혹시 써보신분?

Comment (soeque1) [reply src/model/net.py]: docstrings과 dimension 추가하였습니다.

**Review by soeque1 — COMMENTED** [review]
#16 closed

**Comment by ssaru** [comment]
@soeque1 형준님 혹시 PR작업 반영이 다되시면 리뷰어 분들 멘션한번만 걸어주세요!

**Comment by soeque1** [comment]
@ssaru @inmoonlight 코멘트 반영완료!

**Comment by soeque1** [comment]
@inmoonlight @ssaru 특별한 이슈 없으면 merge 하겠습니다.

**Review by inmoonlight — APPROVED** [review]

File: src/model/net.py (hunk 1 above)
Comment (ssaru) [reply src/model/net.py:106]: @soeque1 
mypy를 써봐도 좋을 것 같긴합니다.
다음 기수때 한번 고려해보겠습니다.

**Review by ssaru — COMMENTED** [review]
@soeque1 
너무 늦게 봐드려 죄송합니다.
LGTM입니다!

제가 궁금한 내용은 다른 PR에 남겼습니다.! (scheduler 관련)

**Comment by soeque1** [comment]
> @soeque1
> 너무 늦게 봐드려 죄송합니다.
> LGTM입니다!
> 
> 제가 궁금한 내용은 다른 PR에 남겼습니다.! (scheduler 관련)

@ssaru approve 하셨길래 merge 하였어요 ㅋㅋ 감사합니다.
